[#3449] Fix user profile update

pi-sigma opened the pull request:
Taiga: https://taiga.maykinmedia.nl/project/open-inwoner/us/3449

pi-sigma committed:
fix: improve logging for failed API services in profile edit

The log message for failing API services (E-suite/OpenKlant) during
user profile updates was generic and uninformative, contributing to
confusion when something went wrong. We now include the id of the user
and the failing service in the log.

Refs: https://taiga.maykinmedia.nl/project/open-inwoner/us/3449

swrichards commented on src/open_inwoner/accounts/views/profile.py:
```diff
@@ -266,8 +266,22 @@ def form_valid(self, form):
+            if not openklant_update_ok:
+                failed_services.append("OpenKlant")
+
+            logger.error(
```
Let's use timeline logger here (maybe with a new wrapper, `system_error`, that automatically sets `log_level=logging.ERROR`).

pi-sigma committed:
fix: set empty values from esuite profile update to None

eSuite does not recognize submission of empty values for a field
(e.g. phone number) as deleting the field. We explicitly set empty
fields submitted via the profile edit form to null to make deletion
via eSuite work.

Refs: https://taiga.maykinmedia.nl/project/open-inwoner/task/3460

swrichards commented on src/open_inwoner/utils/logentry.py:
```diff
@@ -171,3 +171,35 @@ def system_action(
     )
+
+
+def system_error(
```
Just a quick note that we could use some `functool.partial` magic in this module to layer the different calls, e.g. `system_error = functools.partial(system_error, log_level=logging.ERROR)` (similar for warning and debug).

pi-sigma replied:
Nice, didn't think of that!

pi-sigma committed:
fix: correct error handling for API failure during profile update

pi-sigma committed:
chore: update changelog

swrichards commented on src/open_inwoner/openklant/services.py:
```diff
@@ -409,9 +409,11 @@ def update_klant_from_user(
+            "telefoonnummer": user.phonenumber if user.phonenumber else None,
+            "telefoonnummerAlternatief": user.phonenumber_alternative
+            if user.phonenumber_alternative
+            else None,
```
```suggestion
            "emailadres": user.email or None,
            "telefoonnummer": user.phonenumber or None,
            "telefoonnummerAlternatief": user.phonenumber_alternative or None,
```

swrichards commented on src/open_inwoner/accounts/views/profile.py:
```diff
@@ -266,9 +272,24 @@ def form_valid(self, form):
+            # Log which service failed and what fields were being updated
+            failed_services = []
+            if not esuite_update_ok:
+                failed_services.append("eSuite")
```
If we just start adding this immediately in the `except` blocks, we can simplify the check on `l.267`: `if failed_services`

swrichards commented on src/open_inwoner/accounts/views/profile.py:
```diff
@@ -251,13 +251,19 @@ def form_valid(self, form):
+                self.update_klant_via_esuite(
+                    {k: form.cleaned_data[k] for k in form.changed_data}, user
+                )
+            except Exception:
```
Let's add `logger.exception` here and below.

swrichards commented on CHANGELOG.rst:
```diff
@@ -41,16 +41,22 @@ Nieuwe features
+* [:taiga-is:`3460`, :pr:`1904`]: Het indienen van lege waarden tijdens het bijwerken van
+  het profiel, wat leidt tot fouten van e, is opgelost.
+* [:taiga-us:`3449`, :pr:`1904`]: Foutbehandeling voor mislukte API-services tijdens het
+  bijwerken van gebruikersprofielen gerepareerd
```
I think we can combine this to a single entry (also note that `3460` was a task, not an issue):

```suggestion
* [:taiga-ta:`3460`, :taiga-us:`3449`, :pr:`1904`]: Het indienen van lege tekstwaarden tijdens het bijwerken van het profiel, wat leidde tot fouten in eSuite klant, is opgelost.
```

swrichards commented on src/open_inwoner/accounts/views/profile.py:
```diff
@@ -247,28 +247,44 @@ def form_valid(self, form):
+                    {k: form.cleaned_data[k] for k in form.changed_data}, user
+                )
+            except Exception:
+                logger.exception("eSuite failed during profile update for %(user)s" % user)
```
I've found it's good practice to not interpolate parameters in the exception string, it makes it harder to dedupe at Sentry level. I would put the user in the `extra` dictionary:


```suggestion
                logger.exception("eSuite failed during profile update", extra={"user":user})
```

swrichards commented on src/open_inwoner/accounts/views/profile.py:
```diff
@@ -247,28 +247,44 @@ def form_valid(self, form):
+                    {k: form.cleaned_data[k] for k in form.changed_data}, user
+                )
+            except Exception:
+                logger.exception("OpenKlant failed during profile update for %(user)s" % user)
```
```suggestion
                logger.exception("OpenKlant failed during profile update", extra={"user": user})
```

swrichards reviewed (changes requested):
Just a final nitpick re exceptions
